cli_tools/gce_vm_image_import/importer/importer.go

```diff
@@ -81,46 +92,83 @@ func (i importer) Run(ctx context.Context) (loggable service.Loggable, err error
 }
 
 func (i *importer) runInflate(ctx context.Context) (err error) {
-	select {
-	case <-ctx.Done():
-		err = ctx.Err()
-	default:
+	return i.runStep(ctx, func(ctx context.Context) error {
+		var err error
 		i.pd, err = i.inflater.inflate(ctx)
-		i.traceLogs = append(i.traceLogs, i.inflater.traceLogs()...)
-	}
-	return err
+		return err
+	}, i.inflater.cancel, i.inflater.traceLogs)
 }
 
 func (i *importer) runProcess(ctx context.Context) (err error) {
-	processor, err := i.processorProvider.provide(i.pd)
+	i.processor, err = i.processorProvider.provide(i.pd)
 	if err != nil {
 		return err
 	}
+	return i.runStep(ctx, func(ctx context.Context) error { return i.processor.process(ctx) }, i.processor.cancel, i.processor.traceLogs)
+}
+
+func (i *importer) runStep(ctx context.Context, step func(context.Context) error, cancel func(string), getTraceLogs func() []string) (err error) {
+	e := make(chan error)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
```

Comment:
I was hoping importer.go would focus on business logic, which will be important when we add more business logic steps. Think:
1. Native VMDK
2. UEFI tagging
3. OS detection
4. Pre-detecting disk size requirements
5. Switching to different translation logic

Reply:
In what way is this current solution blocking the addition of these new steps?

Reply:
Here, the `runStep` method assumes that steps will always have the signature `func(context.Context) error`. Although this is true today, these upcoming projects will challenge that assumption. Consider pre-detection of disk sizes. Using the straight-line design that I originally envisioned, an `inspector` would return its inspection results, and those results would be passed to `inflater.inflate` as a new parameter. 

    diskFile, err := i.inspector.inspect(ctx)

    // handle error

    if err = i.runInflate(ctx, diskFile.size); err != nil {
      return i.buildLoggable(), err
    }

You have any thoughts on what this would look like with your step framework?

Reply:
We already do something similar where inflater passes i.pd to processor. Step doesn't have to be `func(context) err` if we use closures (like currently for inflater).

Reply:
very good point!